src/driver-sinowealth.c

```diff
@@ -0,0 +1,400 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "libratbag-private.h"
+#include "libratbag-hidraw.h"
+
+#define SINOWEALTH_REPORT_ID_CONFIG 0x4
+#define SINOWEALTH_REPORT_ID_CMD 0x5
+#define SINOWEALTH_CMD_FIRMWARE_VERSION 0x1
+#define SINOWEALTH_CMD_GET_CONFIG 0x11
+#define SINOWEALTH_CONFIG_SIZE 520
+
+#define SINOWEALTH_XY_INDEPENDENT 0x80
+
+/* The PC software only goes down to 400, but PMW3360 doesn't care */
+#define SINOWEALTH_DPI_MIN 100
+#define SINOWEALTH_DPI_MAX 12000
+#define SINOWEALTH_DPI_STEP 100
+
+/* other models might have up to eight */
+#define SINOWEALTH_NUM_DPIS 6
+
+typedef struct __attribute__((packed)) {
+	uint8_t r, g, b;
+} RGB8;
+
+enum rgb_effect {
+	RGB_OFF = 0,
+	RGB_GLORIOUS = 0x1,   /* unicorn mode */
+	RGB_SINGLE = 0x2,     /* single constant color */
+	RGB_BREATHING = 0x5,  /* RGB breathing */
+	RGB_BREATHING7 = 0x3, /* breathing with seven colors */
+	RGB_BREATHING1 = 0xa, /* single color breathing */
+	RGB_TAIL = 0x4,       /* idk what this is supposed to be */
+	RGB_RAVE = 0x7,       /* ig */
+	RGB_WAVE = 0x9
+};
+
+struct sinowealth_config_report {
+	uint8_t report_id; /* SINOWEALTH_REPORT_ID_CONFIG */
+	uint8_t command_id;
+	uint8_t unk1;
+	uint8_t config_write;
+	/* always 0 when config is read from device,
+	 * has to be 0x7b when writing config to device
+	 */
+	uint8_t unk2[6];
+	uint8_t config;
+	/* 0x80 - SINOWEALTH_XY_INDEPENDENT */
+	uint8_t dpi_count:4;
+	uint8_t active_dpi:4;
+	uint8_t dpi_enabled;
+	/* bit set: disabled, unset: enabled
+	 * this structure has support for eight DPI slots,
+	 * but the glorious software only exposes six
+	 */
+	uint8_t dpi[16];
+	/* DPI/CPI is encoded in the way the PMW3360 sensor accepts it
+	 * value = (DPI - 100) / 100
+	 * If XY are identical, dpi[0-6] contain the sensitivities,
+	 * while in XY independent mode each entry takes two chars for X and Y.
+	 */
+	RGB8 dpi_color[8];
+
```

Comment:
Can you keep the style consistent here? Sometimes you add a newline between fields, sometimes you don't.

Reply:
Done